### PR TITLE
feat: add basic contents

### PR DESCRIPTION
**Pull request by Berrysoft**
I will construct the basic contents here.

### CODE REVIEW

**Comment by George-Miao** [comment]
Can you reformat so that each line is actually in one line? i.e. don't add single new lines. If it's too long for editor to show, you can try enabling line wrap.

File: src/preface.md [hunk 1]
```diff
@@ -1 +1,79 @@
...
+# A story of `compio`
+
+Long, long time ago, there was a group "Rust 众".
+People there were interested in asynchronous programming, and the newest, hottest technology io-uring.
+They wondered if the most popular async runtime `tokio` could use this technology as the backend.
+
+The `tokio` community had some trials, but the interface of `tokio-uring` was different from the original `tokio` one, for safety reasons.
+Traditionally, `tokio` is based on reactor model,
+while io-uring is a proactor.
+proactor based IO methods require the ownership of the buffers.
```
Comment (George-Miao) [inline src/preface.md:10]:
```suggestion
Proactor-based IO functions require the ownership of the buffers, a completely different API idiom from the current ones.
```

**Comment by Berrysoft** [comment]
I prefer not doing that because, if in the future we need to modify only one word, the whole paragraph would be changed, which makes the diff hard to read.

File: src/preface.md [hunk 2]
```diff
@@ -1 +1,79 @@
...
+Therefore, unless `tokio` finally decided to release 2.0 for the breaking change, it would never use io-uring as first class.
```
Comment (George-Miao) [inline src/preface.md:11]:
```suggestion
Therefore, unless `tokio` is willing to release 2.0 for the breaking change, io-uring will never be a first class citizen for `tokio`.
```

File: src/preface.md [hunk 3]
```diff
@@ -1 +1,79 @@
...
+Wait - IOCP on Windows is also a proactor, right?
+Yes, but no one really cares about that.
+Windows IOCP is neither newest nor hottest.
+It has been in the system kernel for 20 years.
+Before io-uring, it was the only proactor among the popular operating systems.
+Therefore, `tokio` decided to use the *undocumented* AFD driver, which is a satisfying reactor.
+
+Now io-uring is here.
+Unexpectedly, the greatest open source operating system chooses proactor.
+It's not `tokio`'s fault, because it should keep its API stable.
```
Comment (George-Miao) [inline src/preface.md:22]: Delete this line

File: src/preface.md [hunk 4]
```diff
@@ -1 +1,79 @@
...
+Let's try to find some other async runtimes.
+Unfortunately, there was no other runtime, that used io-uring correctly, and was cross-platform.
```
Comment (George-Miao) [inline src/preface.md:24]:
```suggestion
Let's find an async runtime that uses io-uring correctly and be cross-platform. Unfortunately, there's none.
```

File: src/preface.md [hunk 5]
```diff
@@ -1 +1,79 @@
...
+
+Then I came up with an idea: if there is no such runtime, why cannot I author one?
```
Comment (George-Miao) [inline src/preface.md:26]: Don't use first personal POV in docs. Use "Berrysoft" or other name you find appropriate.

```suggestion
That's when Berrysoft came up with the idea: if there is no such runtime, why cannot I author one?
```

File: src/preface.md [hunk 6]
```diff
@@ -1 +1,79 @@
...
+It was a failure, because `tokio` refused to expose `mio` interfaces, and I ended up with a runtime that consumed CPU 100% even there were no tasks to operate.
+However, with the experience and lessons from the failure, I believed that I had the ability to author a new one.
+
+I started with a runtime based on IOCP.
+It should contain a struct `File` to read and write the files.
+Then it was ported to Linux based on io-uring.
+The API of IOCP and io-uring differs, but their natures are the same.
+The API of the new runtime imitated `tokio` and `monoio`.
+When the first example ran correctly on both Windows and Linux, I was very excited.
+I told about the new runtime in the group.
```
Comment (George-Miao) [inline src/preface.md:37]:
```suggestion
I shared the new runtime to the chat group.
```

File: src/preface.md [hunk 7]
```diff
@@ -1 +1,79 @@
...
+I gave this new runtime the name "compio", which means completion-based IO.
+It should pronounce like /'kɔmpaio/ or /'kɔmpio/.
+The name follows the non-existent convention that an async runtime should be named with a suffix "io".
+Later, I also designed a logo for it.
+It is a letter "C" with a plug inside, which means that it is a "socket".
+
+By the project developed, more people showed interest in it.
+Thanks to @George-Miao (Pop)'s so many efforts on `compio`.
+He is a better developer and reviewer than me.
+When I introduced the runtime in the group, he noticed that and decided to port it to Mac.
```
Comment (George-Miao) [inline src/preface.md:49]:
```suggestion
When I introduced the runtime in the group, he noticed it and decided to port it to Mac.
```

File: src/preface.md [hunk 8]
```diff
@@ -1 +1,79 @@
...
+It was not an easy job, because there is no proactor on Mac.
+He had to simulate a proactor with a reactor.
+After his success, the structure of `compio-driver` finally established.
```
Comment (George-Miao) [inline src/preface.md:52]:
```suggestion
After his success, the structure of `compio-driver` was finally established.
```

File: src/preface.md [hunk 9]
```diff
@@ -1 +1,79 @@
...
+
+`compio` had few users at first.
+It only provided mid-level interfaces, e.g., TCP streams.
+If there were an HTTP client based on `compio`, it would be easier for newcomers to try `compio`.
+Thought of that, I started to author `compio-http`, and finally it became `cyper`.
+`cyper` is based on `compio` and `hyper`, providing HTTP client in `cyper` crate, and `axum` adapters in `cyper-axum`.
+
+Before April Fools of 2024, I decided to make some big news.
+I found that the natures of an async runtime and a GUI framework are the same.
+Therefore, it should be possible to write a GUI async runtime, in which IO operations in the same thread wouldn't freeze the GUI.
```
Comment (George-Miao) [inline src/preface.md:62]:
```suggestion
Therefore, it should be possible to write a GUI async runtime, in which IO operations and GUI coexist in one thread without blocking each other.
```

File: src/preface.md [hunk 10]
```diff
@@ -1 +1,79 @@
...
+I started with another new runtime, copying code from `compio`, and added some other stuffs to handle the message queue on Windows.
```
Comment (George-Miao) [inline src/preface.md:63]:
```suggestion
I started with another new runtime, copying code from `compio`, and added some other stuff to handle the message queue on Windows.
```

**Review by George-Miao — CHANGES_REQUESTED** [review]

**Comment by George-Miao** [comment]
> I prefer not doing that because, if in the future we need to modify only one word, the whole paragraph would be changed, which makes the diff hard to read.

It won't. Diff will correctly label out which exact words are being changed. Current line arrangements look ugly in editor.

**Comment by Berrysoft** [comment]
OK. I'll adjust it, but it takes some time:)

File: src/preface.md (hunk 3 above)
Comment (George-Miao) [inline src/preface.md:22]: Maybe mention a little on how rust's `Future` by-design encourages poll-based API?

**Comment by George-Miao** [comment]
I have copy all content in this crate to compio-website, should we close this? If there's anything I'd like to fix, I will do it directly there.

**Comment by Berrysoft** [comment]
OK, closing it now.

**Pull request closed without merging**